rxstack.1
.TH rxstack 1 "February 1999"
.SH NAME
rxstack \- access the Rexx data queue
.SH SYNOPSIS
.B rxstack
[
.B \-fifo
|
.B \-lifo
] [
.B \-drop
|
.B \-num
|
.B \-peek
|
.B \-pop
|
.B \-print
|
.BI \-string " data"
]
.SH DESCRIPTION
.B rxstack
is a program to access the data queue served by
.BR rxque (1)
and used by
.BR rexx (1).
It can be used to stack the output of a Unix command for use in
Rexx programs or to feed the contents of the queue into a Unix
command.  With no parameters,
.B rxstack
reads lines from the standard input and stacks them in FIFO
(first in, first out) order.  This behaviour may be modified by
supplying an alternative order and/or an alternative action
to perform.
.SH OPTIONS
.TP
.B \-fifo
Specify FIFO order (the first item to be stacked will be the first
item output).
.TP
.B \-lifo
Specify LIFO order (the last item to be stacked will be the first
item output, and this will also be output before all items stacked
in FIFO order).
.TP
.B \-drop
Delete the first item in the queue.
.TP
.B \-num
Print the current number of items in the queue.
.TP
.B \-peek
Print the first item in the queue without changing the queue.
.TP
.B \-pop
Remove and print the first item in the queue.
.TP
.B \-print
Remove and print successive items from the queue until the queue
is exhausted.  (It is possible for this command to leave some
items in the queue if it is terminated by a signal such as SIGPIPE
before the queue is exhausted.)
.TP
.BI \-string " data"
Add the
.I data
to the queue as a single item.
.SH ENVIRONMENT
The environment variable RXSTACK must be set to the name of the
communication socket which is in use by a stack server.
.SH SEE ALSO
.BR rexx (1), " rxque" (1),
The REXX/imc documentation.
.SH AUTHOR
Ian Collier <imc@comlab.ox.ac.uk>


HISTORY
alpha1.2 6 Aug 1992  - initial release

alpha1.3 17 Sep 1992 - fixes the following problems:

   * Installation instructions too brief and not clear enough
   * Error incorrectly reported on the following code format:
       select
           when a
 	     then ...
 	  when b
 	     then ...
       end
   * Clause tracing did not strip leading tabs.   
   * Special variable SIGL not set on calling an internal routine
   * Second parameter of bitxxx() was required; should be optional
   * pos, lastpos, wordpos gave error when the needle was empty
   * sign('-0') = -1 (should be 0)
   * sign() left extra value on calculator stack, causing surprises
   * subword(" to be or not to be ",7) caused crash; similarly for
     word and wordlength
   * symbol() gave incorrect results;
   * translate(x,"") mistook the empty string for an omitted argument
   * wordindex(string,n) gave non-zero when string contained n-1 words
   * x2d() did not recognise second parameter
   * arg() gave "4", not "2" when arguments were: func(1,2,,)
   * binary strings were too restricted (i.e. a multiple of 8 digits and
     no spaces)
   * "DO for=forever" interpreted FOREVER as reserved
   * PROCEDURE EXPOSE used commas as separators instead of spaces
   * date('U') gave wrong day number
   * Leading and trailing spaces were allowed in hex constants
   * Priorities of implicit concatenation, space operator and unary operators
     were wrong
   * random(n,n) returned error; should return n
   * Memory fault occurred after 63 nested PROCEDURE instructions
   * expression after PUSH and QUEUE was required
   * "rxstack" sometimes exited before giving the stack a chance to catch up;
     after that queued() would slowly rise up to its correct value.
   
beta1.4 13 May 1993 - fixes the following problems:

   * Mathematical function name discrepancy between rxmathfn.c and rxfn.c
     now fixed
   * Intermittent "Routine not found" error when using external functions
     now fixed
   * Case of labels is now preserved in "trace l"
   * "parse var a.b" gave "?.B" when a.b undefined, now gives "A.B"
   * "parse var" now signals to NOVALUE when variable is undefined
   * stem. is now different from stem.""
   * "stem.=5; drop stem.5; say stem. stem.5" now says "5 STEM.5" (did say
     "5 5")
   * Control variable in a loop is interpreted on each pass, and END, LEAVE 
     and ITERATE compare symbol names literally.
   * INTERPRET changed to detect incomplete DO/SELECT blocks
   * SELECT (and WHEN) changed to reject multiple statements in WHEN clauses
   * "do name=expri ..." changed to assign "expr + 0" to "name" instead of
     just "expr".
   * UNTIL conditions are tested before incrementing the control variable
   * WHILE/UNTIL checking "modernised" ;-)
   * "procedure expose (varlist)" implemented
   * Fixed uninitialised char* pointer in rxchars2() which made lines()
     sometimes bomb out with segmentation fault
   * Fixed "Unexpected THEN/ELSE" error report for instructions of form
     "otherwise if (cond) then ..."
   * Amended rxmathfn.exec to use single instructions in WHEN clauses
   * Fixed multiplication to say "Arithmetic Overflow" instead of "Invalid
     whole number" when overflow occurs
   * Fixed stacknum() to report error on too-large exponents - it previously
     stacked an erroneous value instead.
   * Fixed sqrt() to work with argument zero
   * Fixed EXIT instruction to record size of new calculator stack
     (previous code occasionally caused hard-to-trace memory crashes)
   * chars(file) calls stat to return the size of file when it is not open
   * chars() now takes into account characters buffered by stdio
   * NUMERIC settings saved across function calls
   * "parse numeric" now works
   * Changed the sqrt() function in rxmathfn.exec
   * Fixed formatting to obey 2*DIGITS rule for small numbers
   * Fixed format() to obey 2*expt rule, also to round the number on input
   
REXX/imc-beta-1.5 4.00 25 Jun 1993 has the following changes:

   * Fixed two erroneous source lines which choke sensitive compilers
   * Distinguished between AIX cc and AIX gcc in the Makefile
   * made "do 3=i" an error instead of a counted loop
   * made "a==4" an error instead of a command
   * Message 15 last word changed from "constant" to "string"
   * '414243'x(3) is now a function call (was abuttal)
   * allowed tab characters in hex and binary strings
   * implemented "drop (list)"
   * Added STUFF_STACK preprocesor symbol
   * Fixed traceback and related things so that, for instance, running
     main.exec, where main, foo and bar are as follows:

     main.exec                   foo.exec                          bar.exec
   
     /* call an OK program */    /* call an erroneous program */   /* error */
     call foo                    say "Entered FOO"                 return a+b
                                 say bar()
                                 say "Leaving FOO"
                                 return
			       
     produces the following output:

     Entered FOO
         2 +++  return a+b
     Error 41 running /tmp/bar.exec, line 2: Bad arithmetic conversion
         3 +++  say bar()
     Error 50 running /tmp/foo.exec, line 3: Error in called routine
         2 +++ call foo
     Error 50 running /tmp/main.exec, line 2: Error in called routine

     whereas only the first four lines would previously have been displayed.
     Similarly, traceback is printed for an error in an external routine
     even when the error is trapped in the calling program.

   * Removed the RC messages from error reports such as the above, because
     they were pointless.
   * Changed the format of RC messages from "RC(n)" to "RC=n"
   * Implemented the third parameter of justify()
   * Corrected "parse version" to show a language level as second token
     instead of the interpreter version.
   * Corrected mtest() to keep the old value if realloc fails

rxstack.c
/* Utility to access the REXX/imc stack server     (C) Ian Collier 1992 */

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<signal.h>
#include<sys/types.h>
#include<sys/socket.h>
#include<unistd.h>

void ex1t();

struct {u_short af;char name[100];}  /* a mimic of struct sockaddr, but */
       sock={AF_UNIX};               /* with a longer name field. */
int socklen;                         /* The length of the socket name */
char *buff;                          /* An input buffer */
unsigned bufflen;                    /* The amount of memory allocated */
int s;                               /* The socket fd */

int main(argc,argv)
int argc;
char **argv;
{
   int fifo=1;                       /* fifo or lifo */
   int c;
   int x2d();
   int pop();
   static char *opts[]={"-fifo","-lifo","-string","-pop","-peek","-drop",
                        "-num","-print"};  /* All the allowed options */
   static int optnum=8;              /* The number of allowed options */
   int opt=0;                        /* position of an option in the above */
   char id;                          /* a command character */
   char *string;                     /* The parameter of -string flag */
   int len;
   static char out[10];              /* A buffer for sprintf */
   char *env=getenv("RXSTACK");      /* The socket's file name */
   if(!env)fputs("RX Stack not found\n",stderr),exit(1);
   if(!(buff=malloc(bufflen=256)))perror("Unable to allocate memory"),exit(1);
   strcpy(sock.name,env);
   socklen=sizeof(u_short)+strlen(env);
   for(c=1;c<argc;c++){              /* Find first "action" flag */
      for(opt=0;opt<optnum&&strcmp(opts[opt],argv[c]);opt++);
      if(opt==optnum)fprintf(stderr,"Invalid option: %s\n",argv[c]),exit(1);
      if(opt>1)break;
      if(opt==0)fifo=1;else fifo=0;  /* Deal with -fifo and -lifo and go on */
   }
   if(argc>++c+(opt==2))fputs("Too many options\n",stderr),exit(1);
   if(opt<2)opt=0;                   /* no "action"; stack from stdin */
   s=socket(AF_UNIX,SOCK_STREAM,0);  /* Try to connect... */
   if(s<0)perror("Error in socket call"),exit(1);
   if(connect(s,(struct sockaddr *)&sock,socklen)<0)perror("Error in connect call"),exit(1);
   id=fifo?'Q':'S';                  /* Queue if fifo, Stack if lifo. */
   switch(opt){                      /* Select the appropriate action */
      case 0:while(1){               /* Default: collect lines from stdin */
                len=0;
                while((c=getchar())!=EOF&&c!='\n'){
                   if(len>=bufflen&&!(buff=realloc(buff,bufflen<<=1)))
                      perror("Out of memory"),exit(1);
                   buff[len++]=c;
                }
                if(c==EOF&&!len)break;
                sprintf(out,"%c%06X\n",id,len);
                if(write(s,out,8)<8||write(s,buff,len)<len)
                   perror("Error writing data"),exit(1);
             }
             break;
      case 2:string=(c>=argc?"":argv[c]);/* Stack argument following -string */
             sprintf(out,"%c%06X\n",id,len=strlen(string)); 
             if(write(s,out,8)<8||write(s,string,len)<len)
                perror("Error writing data"),exit(1);
             break;
      case 3:pop('G',1);break;       /* -pop: use the "Get" command */
      case 4:pop('P',1);break;       /* -peek: use the "Peek" command */
      case 5:if(write(s,"D",1)<1)    /* -drop: write a "Drop" command */
                perror("Error writing data"),exit(1);
             break;
      case 6:if(write(s,"N",1)<1)    /* -num: write a "Num" command; get ans */
                perror("Error reading data"),exit(1);
             if(read(s,buff,7)<7)perror("Error reading data"),exit(1);
             len=x2d(buff);
             printf("%d\n",len);
             fflush(stdout);
             break;
      case 7:while(pop('P',0)!=0xffffff) /* -print: continually peek, then */
                if(write(s,"D",1)<1)     /* drop.                          */
		   perror("Error writing data"),exit(1);
   }               /* This way, if the output is being read by a program which
                      terminates, a SIGPIPE arrives and any unread data is
		      still on the stack. */
   if(opt<3||opt==5){ /* wait for the stack process to catch up */
      signal(15,ex1t); /* kludge to make a zero return code */
      sprintf(out,"K%06X\n\017",getpid()); /* kill me with SIGTERM */
      if(write(s,out,9)<9)perror("Error writing data"),exit(1);
      pause();
   }
   close(s);
   return 0;
}

void ex1t() {close(s);exit(0);} /* on SIGTERM, just exit with return code 0 */

int x2d(hex)       /* Convert six hex digits from "hex" into an integer */
char *hex;
{
   int i,len=0;
   char c;
   for(i=0;i<6;i++){
      len<<=4;
      if((c=hex[i])>='0'&&c<='9')len+=c-'0';
      else if(c>='A'&&c<='F')len+=c-'A'+10;
      else if(c>='a'&&c<='f')len+=c-'a'+10;
   }
   return len;
}

int pop(id,nullok)  /* Print the top value from the stack, using "id" as */
char id;            /* the command character.  If the stack is empty and */
int nullok;         /* nullok!=0 then print a newline, else just return. */
{                   /* The return value is the length given by the stack */
   int len;
   int rdlen;
   int i;
   if(write(s,&id,1)<1)    /* write the command character */
      perror("Error writing data"),exit(1);
   if(read(s,buff,7)<7)    /* Expect a length to be returned */
      perror("Error reading data"),exit(1);
   len=x2d(buff);
   if(len!=0xffffff){      /* Stack wasn't empty, so expect some data */
      if(len>bufflen&&!(buff=realloc(buff,(unsigned)len)))
         perror("Out of memory"),exit(1);
      rdlen=0;
      while(rdlen<len)     /* Read the data */
         if((i=read(s,buff+rdlen,len-rdlen))<1)
	    perror("Error reading data"),exit(1);
	 else rdlen+=i;
      fwrite(buff,1,len,stdout); /* Write the data to stdout */
      putchar('\n');
      fflush(stdout);      /* Make sure the application reads this */
   }
   else if(nullok){        /* Stack was empty; write an empty line */
      putchar('\n');
      fflush(stdout);
   }
   return len;
}
